posthog/tasks/warehouse.py

```diff
@@ -0,0 +1,203 @@
+            resource.save()
+
+    else:
+        resource.status = job["status"]
+        resource.save()
+
+
+DEFAULT_USAGE_LIMIT = 1000000
+ROWS_PER_DOLLAR = 66666  # 1 million rows per $15
+
+
+@app.task(ignore_result=True, max_retries=2)
+def check_external_data_source_billing_limit_by_team(team_id):
+    from posthog.warehouse.external_data_source.connection import deactivate_connection_by_id, activate_connection_by_id
+
+    team = Team.objects.get(pk=team_id)
+    all_active_connections = ExternalDataSource.objects.filter(team=team, status__in=["running", "succeeded"])
+    all_inactive_connections = ExternalDataSource.objects.filter(team=team, status="inactive")
+
+    _usage_limit = _get_data_warehouse_usage_limit(team_id)
+
+    # TODO: consider more boundaries
+    if _usage_limit and team.external_data_workspace_rows_synced_in_month >= (_usage_limit * ROWS_PER_DOLLAR):
+        for connection in all_active_connections:
+            deactivate_connection_by_id(connection.connection_id)
+            connection.status = "inactive"
+            connection.save()
+    else:
+        for connection in all_inactive_connections:
+            activate_connection_by_id(connection.connection_id)
+            connection.status = "running"
+            connection.save()
+
+
+def _get_data_warehouse_usage_limit(team_id):
+    from posthog.cloud_utils import get_cached_instance_license
+    from ee.billing.billing_manager import BillingManager
+
+    license = get_cached_instance_license()
+    team = Team.objects.get(pk=team_id)
+    org = team.organization
+
+    response = BillingManager(license).get_billing(org, None)
+    try:
+        data_warehouse_product = [product for product in response["products"] if product["name"] == "Data Warehouse"][0]
+        usage_limit = data_warehouse_product["usage_limit"]
+    except Exception as e:
+        logger.exception("Data Warehouse: Failed to retrieve data warehouse billing product", exc_info=e)
+        usage_limit = DEFAULT_USAGE_LIMIT
+
+    return usage_limit
+
+
+@app.task(ignore_result=True, max_retries=2)
+def calculate_workspace_rows_synced_by_team(team_id):
+    ph_client = get_ph_client()
+    team = Team.objects.get(pk=team_id)
+    now = datetime.datetime.now(datetime.timezone.utc)
+    begin = team.external_data_workspace_last_synced or now.replace(day=1, hour=0, minute=0, second=0, microsecond=0)
+    end = now
+
+    params = {
+        "workspaceIds": team.external_data_workspace_id,
+        "limit": 100,
+        "offset": 0,
+        "status": "succeeded",
+        "orderBy": "createdAt|ASC",
+        "updatedAtStart": begin.strftime("%Y-%m-%dT%H:%M:%SZ"),
+        "updatedAtEnd": end.strftime("%Y-%m-%dT%H:%M:%SZ"),
+    }
+    result_totals = _traverse_jobs_by_field(ph_client, team, AIRBYTE_JOBS_URL + "?" + urlencode(params), "rowsSynced")
+
+    # reset accumulated to new period if the month has changed
+    if end.month != begin.month:
```

Comment:
this is only possible when team.external_data_workspace_last_synced exists right? not sure when the second condition comes in.

team.external_data_workspace_last_synced and team.external_data_workspace_last_synced.month != begin.month -> this seems impossible, or am I missing something?

Reply:
yeah you're right! i had the test mocked in an incorrect way and confused myself because it was setup in a way that violated what should be expected from airbyte's API